dlup/experimental_backends/__init__.py

```diff
@@ -0,0 +1,133 @@
+# coding=utf-8
+# Copyright (c) dlup contributors
+import os
+import pathlib
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import Callable
+
+import numpy as np
+import openslide
+import PIL
+import pyvips
+import tifffile
+
+from dlup import UnsupportedSlideError
+
+from ._openslide import OpenSlideSlide
+from ._pyvips import PyVipsSlide
+from ._tifffile import TifffileSlide
+from .common import AbstractSlideBackend
+
+
+@lru_cache(maxsize=None)
+def autodetect_backend(filename: os.PathLike) -> AbstractSlideBackend:
+    """
+    Try to read the file in consecutive order of pyvips, openslide, tifffile,
+    by trying to a tile at the lowest resolution.
+    The results are cached, so the test is only performed once per filename.
+
+    Parameters
+    ----------
+    filename
+
+    Returns
+    -------
+    AbstractSlideBackend
+        A backend which is able to read the image.
+    """
+    # Can try to be a bit more complex by first checking the path
+    filename = pathlib.Path(filename)
+    # OpenSlide cannot parse the mpp of tiff's
+    if filename.suffix in [".tif", ".tiff"]:
+        try:
+            return _try_pyvips(filename)
+        except UnsupportedSlideError:
+            pass
+        try:
+            return _try_tifffile(filename)
+        except UnsupportedSlideError:
+            raise UnsupportedSlideError(f"Cannot read {filename} with pyvips or tifffile.")
+
+    try:
```

Comment:
pyvips seems to already "try" to load via openslide. Maybe just use pyvips?

Reply:
Yeah but it can depend on another version.